### ISSUE TITLE
Query result container issues in Query spec

### ISSUE THREAD

**Issue by jamsden** [jamsden/oslc-core-test#21]
The description of the query result could do with at least one example.

The current specification does not describe any container, but related guidance documents do. The current position is that an oslc:QueryCapability may reference Zero-or-one oslc:resourceShape. If specified, that resource shape is the resource shape of the query result container, and it MUST denote at least one OSLC property as being a member property using the oslc:isMemberProperty. That predicate is defined in OSLC Core 3.0 Resource shapes, but was not described in the query spec. If a query capability does not reference a resource shape, then the current \[undocumented\] behaviour is to use rdfs:member.

Some implementations also provide a resource shape for the query result members by providing an oslc:valueShape for the OSLC property that has oslc:isMemberProperty set to true. However, the semantics of that referenced shape is ambiguous - see [https://issues.oasis-open.org/browse/OSLCCORE-145](https://issues.oasis-open.org/browse/OSLCCORE-145).

In many cases, both OSLC Core and OSLC Configuration Management provide other containers, and these are specified as LDPCs. These LDPCs have some similarity with a query base except that they do not support oslc.where or oslc.searcdhTerms. We should align the query result container with LDPCs. This has the advantage that it‘s backwards compatible with OSLC 2.0 query, but has the usual semantics of LDPCs widely used in OSLC Core 3.0.

**Comment by ndjc** [comment]
Works for me.

**Comment by DavidJHoney** [comment]
The query capability resource shape describes the queryable resources that are being queried. It doesn‘t describe the query container result. I have seen several different implementations, some of which include a container of sorts, and at least one that doesn‘t.

**Comment by DavidJHoney** [comment]
Re: "The Query Capability should provide a Resource Shape that describes the query base URI."  
What this means is that the shape describes the resource whose subject URI is the query base URI, which is the query result container.

Re: "Regarding the value shape that describes the members, isn‘t that the purpose of QueryCapability::resourceType? How would this relate to the valueShape of the query capability resourceShape?"  
The query capabilities oslc:resourceType property is described as: "The expected resource type URI that will be returned with this query capability. These would be the URIs found in the result resource‘s rdf:type property. ".  
For example, a query capability for configurations for configurations might have oslc:ResourceType oslc\\\_config:Configuration. However, that says nothing about the resource shape of those configurations and does not provide a means for a client discover such a resource shape. So oslc:resourceType is orthogonal to oslc:resourceShape. The way that a client consuming a query capability discovers the shape of the members is to get the oslc:QueryCapability oslc:resourceShape, then determine the OSLC property describing membership (oslc:isMemberProperty "true"^^xsd:boolean), and then perform a GET on that property‘s oslc:valueShape. It‘s a pity that this requires two separate GETs. If OSLC defined another predicate, say oslc:memberResourceShape, the oslc:QueryCapability could have used that so that a client could GET that shape without having to get the query base resource shape. I don‘t think it‘s worth changing that now. Existing servers that care about the shape of the members returned by the query would already be using the existing representation.

Given that we view backwards compatibility as one of the most important requirements for OSLC 3.0, at a minimum we must preserve the existing semantics:  
1) If a oslc:QueryCapability does not reference a resource shape, the default membership predicate is rdfs:member.  
2) If a oslc:QueryCapability references a resource shape, that shape must include at least one OSLC property with oslc:isMemberProperty="true"^^xsd:boolean.

If we want consistency with both 3.0 LDPC usage and LDPCs that are typically like query capabilities but without support for oslc.where, then we also need the query response to:  
3) Be an LDPC  
4) Have the required LDPC headers in the response.

That means for 3.0 usage compatible with 2.0 usage, we need to satsify all 4 requirements. My proposal does that. If a 2.0 or 3.0 server only chooses to satsify #1 and #2, then consumers should still work. The examples show that this is easy to implement, and if a server uses ldp:Contains, the query result keeps a compact and easyily understandable container representation.

**Comment by ndjc** [comment]
Query Syntax 2.0 states that:

   _"The members of the container are defined by one or more multi-valued membership properties of the starting subject resource. The membership properties are given by the the properties described as membership properties (via oslc:isMemberProperty) in the Resource Shape of the starting subject resource."_

This implies there can be more than one occurrence with isMemberProperty=true. The current draft states that "The resource shape referenced by oslc:resourceShape _must_ declare exactly one OSLC property with oslc:isMemberProperty "true"^^xsd:boolean".

This would break compatibility with existing OSLC 2 servers that returned 2 or more member properties. Also note the strength of the MUST in the 3.0 draft is weakened anyway by two surrounding MAYs - the shape may or may not exist in the first place, and the member property may or may not have a value shape.

**Comment by DavidJHoney** [comment]
I disagree with deprecating it. Our primary goal is to maintain backwards compatibility with 2.0 clients. 3.0 server implementations that don‘t use oslc:isMemberProperty are unlikely to work with 2.0 clients. The current committed spec version provides clearer guidance on this, and achieves the two goals of maintaining 2.0 client compatibilty, and supporting LPDCs that are consistent with other query-like LDPCs that are are not actually query capabilities.

**Comment by DavidJHoney** [comment]
It‘s also woth noting that the old and obsolete OSLC Simple Query document referenced in the original OSLC Query spec described some container semantics. But it appears that none of that made it into the OSLC Query 2.0 specification.

One possible view one might take is that since OSLC Query 2.0 did not describe a result container, we might include LDP support with a MAY rather than MUST, and state that consumers SHOULD not rely on there being such a result container.

**Comment by DavidJHoney** [comment]
Proposal to address Nick‘s comment above:  
Change "The resource shape referenced by oslc:resourceShape _must_ declare exactly one OSLC property with oslc:isMemberProperty "true"^xsd:boolean" to "The resource shape referenced by oslc:resourceShape _must_ declare at least one OSLC property with oslc:isMemberProperty "true"^xsd:boolean"
